src/datadog/platform_util.cpp

```diff
@@ -281,5 +286,126 @@ Expected<InMemoryFile> InMemoryFile::make(StringView) {
 }
 #endif
 
+namespace container {
+namespace {
+#if defined(__linux__) || defined(__unix__)
+/// Magic numbers from linux/magic.h:
+/// <https://github.com/torvalds/linux/blob/ca91b9500108d4cf083a635c2e11c884d5dd20ea/include/uapi/linux/magic.h#L71>
+constexpr uint64_t CGROUP_SUPER_MAGIC = 0x27e0eb;
+constexpr uint64_t CGROUP2_SUPER_MAGIC = 0x63677270;
+
+/// Magic number from linux/proc_ns.h:
+/// <https://github.com/torvalds/linux/blob/5859a2b1991101d6b978f3feb5325dad39421f29/include/linux/proc_ns.h#L41-L49>
+constexpr ino_t HOST_CGROUP_NAMESPACE_INODE = 0xeffffffb;
+
+/// Represents the cgroup version of the current process.
+enum class Cgroup : char { v1, v2 };
+
+Optional<ino_t> get_inode(std::string_view path) {
+  struct stat buf;
+  if (stat(path.data(), &buf) != 0) {
+    return nullopt;
+  }
+
+  return buf.st_ino;
+}
+
+// Host namespace inode number are hardcoded, which allows for dectection of
+// whether the binary is running in host or not. However, it does not work when
+// running in a Docker in Docker environment.
+bool is_running_in_host_namespace() {
+  if (auto inode = get_inode("/proc/self/ns/cgroup")) {
+    return *inode == HOST_CGROUP_NAMESPACE_INODE;
+  }
+
+  return false;
+}
+
+Optional<Cgroup> get_cgroup_version() {
+  struct statfs buf;
+
+  if (statfs("/sys/fs/cgroup", &buf) != 0) {
+    return nullopt;
+  }
+
+  if (buf.f_type == CGROUP_SUPER_MAGIC)
+    return Cgroup::v1;
+  else if (buf.f_type == CGROUP2_SUPER_MAGIC)
+    return Cgroup::v2;
+
+  return nullopt;
+}
+
+Optional<std::string> find_docker_container_id_from_cgroup() {
+  constexpr std::string_view cgroup_path = "/proc/self/cgroup";
```

Comment:
Do we gain something by not directly writing the `"/proc/self/cgroup" in the ifstream parameters ? I'm totally OK with staying like that though.

Reply:
not really, it's a relic of an old implementation.